apps/labs/posts/numpy-python-api-cleanup.md

```diff
@@ -0,0 +1,345 @@
+* Local exports - members available only from the `numpy.lib` or `numpy.lib.<submodule>`.
+
+Members exported to the main namespace ended up in private files, such as
+`numpy.lib._array_utils_impl`, where ones exported locally received dedicated
+submodules, e.g. `numpy.lib.array_utils`.
+
+As a result, the number of members in `numpy.lib` reduced from:
+
+```python
+>>> import numpy as np
+>>> len(dir(np.lib))
+206
+```
+
+To:
+
+```python
+>>> import numpy as np
+>>> len(dir(np.lib))
+46
+```
+
+Now `numpy.lib` hosts only a handful of functions and submodules with well-defined
+purposes.
+
+### Renaming `numpy.core` to `numpy._core`
+
+In my opinion, the most challenging task was renaming the `numpy.core` submodule to
+`numpy._core`. It concerned most of the NumPy's source code and affected C-level and
+downstream libraries as well.
+
+In terms of C-level code, it is worth emphasizing that not only does Python use C
+functions through compiled extension modules, but also C code imports NumPy's Python API
+via Python's C-API!
+
+For the source code that is compiled into shared objects and extension modules, we only
+had to rename `core` imports to `_core`. A more complex issue appeared for header files
+that are included in third party objects. We wanted to make sure that library/executable
+compiled with numpy 2.0 will work with numpy 1.x installed (and vice versa). For this
+purpose, we used two simple mechanisms:
+
+|                      | Built with numpy 1.x   | Built with numpy 2.0            |
+|----------------------|------------------------|---------------------------------|
+| numpy 1.x installed  | standard execution     | falls back to second C's import |
+| numpy 2.0 installed  | uses `numpy.core` stub | standard execution              |
+
+**Stubs** - After renaming `numpy.core` submodule to `numpy._core` we did not completely
+remove the former. Instead, we replaced it with a stub module that replicates
+`numpy._core` by importing it, but also generates an appropriate warning when accessed.
+This was necessary to ensure that objects compiled with numpy 1.x will continue to work
+with the numpy 2.0 release.
+
+**Fallback imports** - An object compiled with numpy 2.0 headers should anticipate
+different versions of locally available NumPy in the runtime. Therefore, a simple
+"import fallback" mechanism has been implemented to cover both cases:
+
+```c
+PyObject *numpy = PyImport_ImportModule("numpy._core._multiarray_umath");
+if (numpy == NULL && PyErr_ExceptionMatches(PyExc_ModuleNotFoundError)) {
+  PyErr_Clear();
+  numpy = PyImport_ImportModule("numpy.core._multiarray_umath");
+}
+
+if (numpy == NULL) {
+  PyErr_SetString(PyExc_ImportError, "_multiarray_umath failed to import");
+  return -1;
+}
+```
+
+Other libraries have also followed our lead and started working on renaming `core`
+submodule to `_core`, such as Pandas - [(Pandas PR)](https://github.com/pandas-dev/pandas/pull/55429).
+
```

Comment:
I'd also discuss the headaches for making old pickles loadable in this section

Reply:
Done